## application mobile pour la dépendance à la loterie

Dans ce projet, nous allons contribuer au développement d'une application mobile en écrivant quelques fonctions principalement axées sur le calcul des probabilités. L'application vise à la fois à prévenir et à traiter la dépendance à la loterie en aidant les gens à mieux estimer leurs chances de gagner.

L'idée de l'application vient d'un institut médical spécialisé dans le traitement des dépendances au jeu. L'institut dispose déjà d'une équipe d'ingénieurs qui construira l'application, mais ils ont besoin de nous pour créer le noyau logique de l'application et calculer les probabilités. Pour la première version de l'application, ils veulent que nous nous concentrions sur la loterie 6/49 et que nous construisions des fonctions capables de répondre aux questions suivantes:

 -   Quelle est la probabilité de gagner le gros lot avec un seul billet?
 -   Quelle est la probabilité de gagner le gros lot si nous jouons 40 tickets différents (ou tout autre nombre)?
 -   Quelle est la probabilité d'avoir au moins cinq (ou quatre, ou trois) numéros gagnants sur un seul billet?

Le scénario que nous suivons tout au long de ce projet est fictif - l'objectif principal est de pratiquer l'application des concepts de probabilité et de combinatoire (permutations et combinaisons) dans un cadre qui simule un scénario du monde réel.

### Fonctions principales

Ci-dessous, nous allons écrire deux fonctions que nous utiliserons fréquemment:

   -  factorial() - une fonction qui calcule les factorielles
   -  combinations() - une fonction qui calcule les combinaisons

In [1]:
def factorial(n):
    result = 1
    for i in range(1,n+1):
        result *= i
    return result

def combinations(n, k):
    return factorial(n)/(factorial(k)*factorial(n-k))

### Probabilité avec un ticket

Nous devons créer une fonction qui calcule la probabilité de gagner le gros prix pour un ticket donné. Pour chaque tirage, six numéros sont tirés d'un ensemble de 49, et un joueur gagne le gros lot si les six numéros sur leurs billets correspondent à tous les six numéros.

L'équipe d'ingénieurs nous a dit que nous devons être conscients des détails suivants lorsque nous écrivons la fonction:

   -  Dans l'application, l'utilisateur entre six nombres différents de 1 à 49.
   -  Sous le capot, les six nombres viendront sous forme de liste Python et serviront d'entrée à notre fonction.
   -  L'équipe d'ingénierie souhaite que la fonction imprime la valeur de probabilité de manière amicale - d'une manière que les personnes sans formation de probabilité soient capables de comprendre.

Ci-dessous, nous écrivons la fonction one_ticket_probability(), qui prend une liste de six nombres uniques et imprime la probabilité de gagner d'une manière facile à comprendre.

In [2]:
def one_ticket_probability(liste):
    total_outcomes = combinations(49,6)
    prob_1_ticket = 1/total_outcomes
    perc_1_ticket = prob_1_ticket * 100
    print('''vous avez {: .7f}% de probabilité de gagner à la loterie avec votre combinaison {}, {}, {}, {}, {}, {}.
           En d'autres termes, vous avez 1 sur {:,} chances de gagner.'''.format(perc_1_ticket,
                                                                                 liste[0], liste[1], liste[2],
                                                                                 liste[3], liste[4], liste[5],
                                                                                 int(total_outcomes)))

In [3]:
a = [1,2,3,4,5,6]
one_ticket_probability(a)

vous avez  0.0000072% de probabilité de gagner à la loterie avec votre combinaison 1, 2, 3, 4, 5, 6.
           En d'autres termes, vous avez 1 sur 13,983,816 chances de gagner.


### Vérification des données historiques pour la loterie du Canada

L'institut souhaite également que nous considérions les données provenant du jeu de loterie national 6/49 au Canada. L'ensemble de données contient des données historiques pour 3665 dessins, datant de 1982 à 2018 (l'ensemble de données peut être téléchargé [ici](https://www.kaggle.com/datascienceai/lottery-dataset)).

In [4]:
import pandas as pd
lottery_canada = pd.read_csv('OneDrive\Documents\my_datasets\L649.csv')
lottery_canada.shape

(3665, 11)

In [5]:
lottery_canada.head(3)

PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
0      649            1                0  6/12/1982               3   
1      649            2                0  6/19/1982               8   
2      649            3                0  6/26/1982               1   

   NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
0              11              12              14              41   
1              33              36              37              39   
2               6              23              24              27   

   NUMBER DRAWN 6  BONUS NUMBER  
0              43            13  
1              41             9  
2              39            34

In [6]:
lottery_canada.tail(3)

PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
3662      649         3589                0  6/13/2018               6   
3663      649         3590                0  6/16/2018               2   
3664      649         3591                0  6/20/2018              14   

      NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
3662              22              24              31              32   
3663              15              21              31              38   
3664              24              31              35              37   

      NUMBER DRAWN 6  BONUS NUMBER  
3662              34            16  
3663              49             8  
3664              48            17

## Fonction de vérification des données historiques¶

L'équipe d'ingénierie nous dit que nous devons écrire une fonction qui peut aider les utilisateurs à déterminer s'ils auraient déjà gagné en utilisant une certaine combinaison de six nombres. Voici les détails dont nous devons être conscients:

  -   Dans l'application, l'utilisateur entre six nombres différents de 1 à 49.
  -   Sous le capot, les six nombres viendront sous forme de liste Python et serviront d'entrée à notre fonction.
  -   L'équipe d'ingénierie veut que nous écrivions une fonction qui imprime le nombre de fois où la combinaison sélectionnée s'est produite.

Nous allons commencer par extraire tous les numéros gagnants de l'ensemble de données de loterie. La fonction extract_numbers() parcourra chaque ligne du dataframe et extraira les six numéros gagnants sous la forme d'un ensemble Python.

In [7]:
def extract_numbers(row):
    row = row[4:10]
    row = set(row.values)
    return row

winning_numbers = lottery_canada.apply(extract_numbers, axis=1)
winning_numbers.head()

0    {3, 41, 11, 12, 43, 14}
1    {33, 36, 37, 39, 8, 41}
2     {1, 6, 39, 23, 24, 27}
3     {3, 9, 10, 43, 13, 20}
4    {34, 5, 14, 47, 21, 31}
dtype: object

Ci-dessous, nous écrivons la fonction check_historical_occurrence() qui prend en compte les numéros d'utilisateur et les numéros historiques et imprime des informations concernant le nombre d'occurrences.

In [8]:
def check_historical_occurrence(user_numbers, historical_numbers):   
    
    user_numbers_set = set(user_numbers)
    check_occurrence = historical_numbers == user_numbers_set
    n_occurrences = check_occurrence.sum()
    
    if n_occurrences == 0:
        print('''La combinaison {} ne s'est jamais produite.'''.format(user_numbers))
        
    else:
        print('''Le nombre de fois que la combinaison {} s'est produite dans le passé est {}.'''.format(user_numbers, n_occurrences))

In [9]:
test_input_3 = [33, 36, 37, 39, 8, 41]
check_historical_occurrence(test_input_3, winning_numbers)

Le nombre de fois que la combinaison [33, 36, 37, 39, 8, 41] s'est produite dans le passé est 1.


In [10]:
test_input_4 = [3, 2, 44, 22, 1, 44]
check_historical_occurrence(test_input_4, winning_numbers)

La combinaison [3, 2, 44, 22, 1, 44] ne s'est jamais produite.


## Probabilité multi-ticket

Pour la première version de l'application, les utilisateurs devraient également être en mesure de trouver la probabilité de gagner s'ils jouent plusieurs tickets différents. Par exemple, quelqu'un peut avoir l'intention de jouer 15 billets différents et il veut connaître la probabilité de gagner le gros lot.

L'équipe d'ingénierie souhaite que nous soyons conscients des détails suivants lorsque nous écrivons la fonction:

  -   L'utilisateur saisira le nombre de tickets différents auxquels il souhaite jouer (sans saisir les combinaisons spécifiques qu'il souhaite jouer).
  -   Notre fonction verra un entier entre 1 et 13 983 816 (le nombre maximum de tickets différents).
  -   La fonction doit imprimer des informations sur la probabilité de gagner le gros prix en fonction du nombre de tickets différents joués.

La fonction multi_ticket_probability() ci-dessous prend le nombre de tickets et imprime les informations de probabilité en fonction de l'entrée.

In [11]:
def multi_ticket_probability(n_tickets):
    
    n_combinations = combinations(49, 6)
    
    probability = n_tickets / n_combinations
    percentage_form = probability * 100
    
    if n_tickets == 1:
        print('''Vos chances de gagner le gros lot avec un seul billet sont de {: .6f}%.
En d'autres termes, vous avez 1 sur {:,} chances de gagner.'''.format(percentage_form, int(n_combinations)))
    
    else:
        combinations_simplified = round(n_combinations / n_tickets)   
        print('''Vos chances de gagner le gros lot avec {:,} différents billets sont de {: .6f}%.
En d'autres termes, vous avez 1 sur {:,} chances de gagner.'''.format(n_tickets, percentage_form,
                                                               combinations_simplified))

In [12]:
test_inputs = [1, 10, 100, 10000, 1000000, 6991908, 13983816]

for test_input in test_inputs:
    multi_ticket_probability(test_input)
    print('------------------------')

Vos chances de gagner le gros lot avec un seul billet sont de  0.000007%.
En d'autres termes, vous avez 1 sur 13,983,816 chances de gagner.
------------------------
Vos chances de gagner le gros lot avec 10 différents billets sont de  0.000072%.
En d'autres termes, vous avez 1 sur 1,398,382 chances de gagner.
------------------------
Vos chances de gagner le gros lot avec 100 différents billets sont de  0.000715%.
En d'autres termes, vous avez 1 sur 139,838 chances de gagner.
------------------------
Vos chances de gagner le gros lot avec 10,000 différents billets sont de  0.071511%.
En d'autres termes, vous avez 1 sur 1,398 chances de gagner.
------------------------
Vos chances de gagner le gros lot avec 1,000,000 différents billets sont de  7.151124%.
En d'autres termes, vous avez 1 sur 14 chances de gagner.
------------------------
Vos chances de gagner le gros lot avec 6,991,908 différents billets sont de  50.000000%.
En d'autres termes, vous avez 1 sur 2 chances de gagner.
------

## Moins de numéros gagnants

Dans la plupart des loteries 6/49, il y a des prix plus petits si le billet d'un joueur correspond à deux, trois, quatre ou cinq des six numéros tirés. Cela signifie que les joueurs pourraient être intéressés à découvrir la probabilité d'avoir deux, trois, quatre ou cinq numéros gagnants - pour la première version de l'application, les utilisateurs devraient être en mesure de trouver ces probabilités.

Voici les détails dont nous devons être conscients lorsque nous écrivons une fonction pour rendre les calculs de ces probabilités possibles:

  -  Dans l'application, l'utilisateur saisit:
      -  six nombres différents de 1 à 49; et
      -  un entier entre 2 et 5 qui représente le nombre de numéros gagnants attendus
  -  Notre fonction imprime des informations sur la probabilité d'avoir un certain nombre de numéros gagnants

Pour calculer les probabilités, nous disons à l'équipe d'ingénierie que la combinaison spécifique sur le ticket n'est pas pertinente et que nous n'avons besoin que de l'entier entre 2 et 5 représentant le nombre de numéros gagnants attendus. Par conséquent, nous écrirons une fonction nommée probabilité_less_6() qui prend un entier et affiche des informations sur les chances de gagner en fonction de la valeur de cet entier.

La fonction ci-dessous calcule la probabilité que le ticket d'un joueur corresponde exactement au nombre donné de numéros gagnants. Si le joueur veut connaître la probabilité d'avoir cinq numéros gagnants, la fonction retournera la probabilité d'avoir exactement cinq numéros gagnants (ni plus ni moins). La fonction ne renvoie pas la probabilité d'avoir au moins cinq numéros gagnants.

In [13]:
def probability_less_6(n_winning_numbers):
    
    n_combinations_ticket = combinations(6, n_winning_numbers)
    n_combinations_remaining = combinations(43, 6 - n_winning_numbers)
    successful_outcomes = n_combinations_ticket * n_combinations_remaining
    
    n_combinations_total = combinations(49, 6)    
    probability = successful_outcomes / n_combinations_total
    
    probability_percentage = probability * 100    
    combinations_simplified = round(n_combinations_total/successful_outcomes)    
    print('''Vos chances d'avoir {} numéros gagnants avec ce billet sont de {: .6f}%.
En d'autres termes, vous avez 1 sur {:,} chances de gagner.'''.format(n_winning_numbers, probability_percentage,
                                                                      int(combinations_simplified)))

In [14]:
for test_input in [2, 3, 4, 5]:
    probability_less_6(test_input)
    print('--------------------------')

Vos chances d'avoir 2 numéros gagnants avec ce billet sont de  13.237803%.
En d'autres termes, vous avez 1 sur 8 chances de gagner.
--------------------------
Vos chances d'avoir 3 numéros gagnants avec ce billet sont de  1.765040%.
En d'autres termes, vous avez 1 sur 57 chances de gagner.
--------------------------
Vos chances d'avoir 4 numéros gagnants avec ce billet sont de  0.096862%.
En d'autres termes, vous avez 1 sur 1,032 chances de gagner.
--------------------------
Vos chances d'avoir 5 numéros gagnants avec ce billet sont de  0.001845%.
En d'autres termes, vous avez 1 sur 54,201 chances de gagner.
--------------------------
